File: src/api/provider-request.ts
import { invoke } from "./client";
import type { ProviderModel } from "./providers";
import type { ResponsesOptions, UpstreamProtocol } from "./shared";

/** The connection under test, independent of saved provider metadata. */
export interface ProviderRequestConnection {
  baseUrl: string;
  apiKey: string;
  upstreamProtocol: UpstreamProtocol;
  responsesOptions: ResponsesOptions | null;
  defaultModel: string | null;
}

export type ProviderRequestTarget =
  | { kind: "saved"; profileId: string }
  | { kind: "draft"; connection: ProviderRequestConnection };

/** Credential-free metadata for one backend-owned, single-use request. */
export interface ProviderRequestPreparation {
  requestId: string;
  endpoint: string;
  upstreamProtocol: UpstreamProtocol;
  defaultModel: string | null;
  prompt: string;
}

export type ProviderRequestOutcome =
  | "success"
  | "authenticationFailed"
  | "rateLimited"
  | "httpError"
  | "networkError"
  | "timeout"
  | "invalidResponse"
  | "cancelled";

export type ProviderDiagnosticKind =
  | "dns"
  | "tls"
  | "network"
  | "timeout"
  | "websocketUnsupported"
  | "endpoint"
  | "authentication"
  | "requestParameters"
  | "modelNotFound"
  | "rateLimit"
  | "upstream"
  | "responseParse"
  | "streamParse";

export interface ProviderDiagnostic {
  kind: ProviderDiagnosticKind;
  endpoint: string;
  transport: "http" | "websocket";
  status: number | null;
  requestId: string | null;
  body: string | null;
  bodyTruncated: boolean;
  message: string;
}

export interface ProviderRequestResult {
  outcome: ProviderRequestOutcome;
  status: number | null;
  latencyMs: number;
  /** The model reported by the provider, never a copy of the requested ID. */
  model: string | null;
  reply: string | null;
  error: string | null;
  diagnostic: ProviderDiagnostic | null;
  at: string;
}

/** The only provider-request command that accepts a connection. A saved
 * profile sends its id; an unsaved draft sends its connection once, and every
 * later command refers to it by the opaque token returned here. */
export function prepareProviderRequest(target: ProviderRequestTarget): Promise<ProviderRequestPreparation> {
  return invoke("prepare_provider_request", { target });
}

/** Models offered by the connection a preparation token stands for. The
 * backend resolves the credential from its prepared target, so no credential
 * travels with the call. */
export function fetchProviderRequestModels(requestId: string): Promise<ProviderModel[]> {
  return invoke<ProviderModel[]>("fetch_provider_request_models", { requestId });
}

export function executeProviderRequest(requestId: string, model: string): Promise<ProviderRequestResult> {
  return invoke("execute_provider_request", { request: { requestId, model } });
}

export function cancelProviderRequest(requestId: string): Promise<boolean> {
  return invoke("cancel_provider_request", { requestId });
}

File: src-tauri/src/commands/provider_request.rs
use super::error::{blocking, state, CommandError};
use crate::provider_request::{
    self, ProviderRequestInput, ProviderRequestPreparation, ProviderRequestResult,
    ProviderRequestTarget, ProviderRequests,
};
use tauri::Manager;

/// The only command that accepts a connection. Saved profiles send just their
/// id; an unsaved draft sends its connection once, and every later command
/// refers to it by the opaque token returned here.
#[tauri::command]
pub async fn prepare_provider_request(
    app: tauri::AppHandle,
    target: ProviderRequestTarget,
) -> Result<ProviderRequestPreparation, CommandError> {
    let local = state(&app)?;
    let requests = app.state::<ProviderRequests>().inner().clone();
    blocking(move || provider_request::prepare(&requests, &local.configuration(), target)).await
}

/// Models offered by the connection a preparation token stands for. The
/// backend resolves the credential from its prepared target, so no credential
/// travels with the command.
#[tauri::command]
pub async fn fetch_provider_request_models(
    app: tauri::AppHandle,
    request_id: String,
) -> Result<Vec<crate::probe::ProviderModel>, CommandError> {
    let local = state(&app)?;
    let requests = app.state::<ProviderRequests>().inner().clone();
    provider_request::fetch_models(requests, local.configuration(), request_id).await
}

#[tauri::command]
pub async fn execute_provider_request(
    app: tauri::AppHandle,
    request: ProviderRequestInput,
) -> Result<ProviderRequestResult, CommandError> {
    let local = state(&app)?;
    let requests = app.state::<ProviderRequests>().inner().clone();
    provider_request::execute(requests, local.configuration(), request).await
}

#[tauri::command]
pub async fn cancel_provider_request(
    app: tauri::AppHandle,
    request_id: String,
) -> Result<bool, CommandError> {
    let requests = app.state::<ProviderRequests>().inner().clone();
    requests.cancel(&request_id).await
}
